firewood/src/hashednode.rs

```diff
@@ -308,28 +308,28 @@ impl HasUpdate for Vec<u8> {
 pub(crate) struct HashPreimage<'a, K: Iterator<Item = u8> + Clone, V: AsRef<[u8]>> {
     pub(crate) key: K,
     pub(crate) value_digest: Option<V>,
-    pub(crate) children: Option<&'a [TrieHash; BranchNode::MAX_CHILDREN]>,
+    pub(crate) children: &'a [Option<(LinearAddress, Option<TrieHash>)>; BranchNode::MAX_CHILDREN],
 }
 
 impl<'a, K: Iterator<Item = u8> + Clone, V: AsRef<[u8]>> HashPreimage<'a, K, V> {
     /// Calls `buf.Update` for each byte in the pre-image of `node`.
     fn write<H: HasUpdate>(mut self, buf: &mut H) {
-        if let Some(children) = self.children {
-            let children_iter = children
-                .iter()
-                .enumerate()
-                .filter(|(_, hash)| **hash != Default::default());
-
-            let num_children = children_iter.clone().count() as u64;
-            add_varint_to_buf(buf, num_children);
-
-            for (index, hash) in children_iter {
-                add_varint_to_buf(buf, index as u64);
-                buf.update(hash);
-            }
-        } else {
-            let num_children: u64 = 0;
-            add_varint_to_buf(buf, num_children);
+        // Iterates over children whose hashes are present.
+        let children_iter = self
+            .children
+            .iter()
+            .enumerate()
+            .filter_map(|(i, child_option)| match child_option {
+                Some((_, Some(hash))) => Some((i, hash)),
+                _ => None,
+            });
+
+        let num_children = children_iter.clone().count() as u64;
```

Comment:
> Yikes. Do you really need to do this?

We need the hash preimage format to be the same as merkledb, so I think we do.

I think the choice we have is either:
1. Iterate over `children_iter` twice by cloning it. This is the current behavior. Note that this iterates at most 16 times.
2. Iterate over `children_iter` once. For each child, increment a counter and write the child to a vec. Then, write the number of children to `buf` and then write the children vec to `buf`. Note we need to maintain this serialization format for compatibility with merkledb.

I figured (1) is probably more performant. Do you think that's right? Am I missing some other alternative?

> When is the buf a hash vs a Vec?

We write the preimage to a hasher when we're actually hashing a node. We write the preimage to a vec when verifying proofs.

Reply:
> In the case where it's a Vec, you can insert a 0 then modify with the proper count after, avoiding the clone.

Ah I missed this before. Yeah I think this is doable when `buf` is a vec. But we do sometimes want to write into a hasher rather than a vec. Not sure how to square that circle.

Reply:
I think it should stay as is. Cloning the iterator is cheap -- it just clones the state information.